core/src/main/java/org/elasticsearch/rest/DeprecationRestChannel.java

```diff
@@ -0,0 +1,130 @@
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.rest;
+
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.io.stream.BytesStreamOutput;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * {@code DeprecationRestChannel} is a proxy for an existing {@link RestChannel} that decorates the {@link RestResponse} with
+ * the {@link #DEPRECATION_HEADER deprecation header} and a warning message to denote that the existing request/response is
+ * deprecated.
+ */
+public class DeprecationRestChannel implements RestChannel {
+    /**
+     * The "Warning" Header comes from RFC-7234. As the RFC describes, it's generally used for caching purposes, but it can be
+     * used for <em>any</em> warning.
+     *
+     * https://tools.ietf.org/html/rfc7234#section-5.5
+     */
+    public static final String DEPRECATION_HEADER = "Warning";
+
+    /**
+     * The proxied canchannel.
+     */
+    private final RestChannel channel;
+    /**
+     * The deprecation message.
+     */
+    private final String warning;
+
+    /**
+     * Create a new {@link DeprecationRestChannel}.
+     *
+     * @param channel The channel to proxy
+     * @param warning The warning to add to the response. This should only be ASCII as it goes into a header!
+     * @throws NullPointerException if any parameter is {@code null}
+     */
+    public DeprecationRestChannel(RestChannel channel, String warning) {
+        assert Strings.validHeaderValue(warning);
+
+        // required
+        this.channel = Objects.requireNonNull(channel);
+        this.warning = Objects.requireNonNull(warning);
+    }
+
+    /**
+     * {@inheritDoc}
+     * <p>
+     * Adds the {@link #DEPRECATION_HEADER} to the {@code response}.
+     *
+     * @param response The response to send to the client
+     */
+    @Override
+    public void sendResponse(RestResponse response) {
+        response.addHeader(DEPRECATION_HEADER, warning);
+
+        // actually send the response
+        channel.sendResponse(response);
+    }
+
+    /**
+     * {@inheritDoc}
```

Comment:
> I disagree with the notion that the reviews make it acceptable to drop it because only the submitter and reviewer(s) are aware of what happened, the reasoning, and any assumptions that goes into the changes.

Your claim was that not having it there means one of two things: quickly written code, or we accept the same contract as the parent. I'm saying if we are writing and reviewing code carefully (which we are, right?), then it can not be the case that it's there because the code was written quickly. At least one of the writer and the reviewer (preferably both) have carefully considered the issue at hand.

> Particularly because of Guice, it's extremely annoying to jump through the code with undocumented-after-undocumented thing, having to dig into the weeds to find things that should be pretty simple.

This has nothing to do with whether or not `inheritdoc` is present.

> This is more in regard to our general disregard for Javadoc, not `inheritdoc`

I don't think we have "disregard" for it so much as we don't do a good job of using it when applicable.

I consider the rules around the inheriting of Javadocs here to be incredibly straightforward that they require no clarification. Do you *explicitly* cast every `int` parameter to a `long` when assigning an `int` to `long`? I don't, because the rules are straightforward.

> Applying inheritdoc also starts a good practice of actually adding Javadoc.

It's just clutter.

> Documentation is supposed to help to give context about the contract without reading the raw implementation (e.g., side-effects, can it ever return `null`, when does it throw exceptions, etc.).

We are not arguing over whether or not documentation should be present, we are arguing about cluttering the code with three lines that add nothing because they are otherwise implied.